test/lib/ansible_test/_internal/containers.py

```diff
@@ -221,6 +235,126 @@ def run_support_container(
     return descriptor
 
 
+def run_container(
+        args: EnvironmentConfig,
+        image: str,
+        name: str,
+        options: t.Optional[list[str]],
+        cmd: t.Optional[list[str]] = None,
+        create_only: bool = False,
+) -> str:
+    """Run a container using the given docker image."""
+    options = list(options or [])
+    cmd = list(cmd or [])
+
+    options.extend(['--name', name])
+
+    network = get_docker_preferred_network_name(args)
+
+    if is_docker_user_defined_network(network):
+        # Only when the network is not the default bridge network.
+        options.extend(['--network', network])
+
+    for _iteration in range(1, 3):
+        try:
+            if create_only:
+                stdout = docker_create(args, image, options, cmd)[0]
+            else:
+                stdout = docker_run(args, image, options, cmd)[0]
+
+            if args.explain:
+                stdout = ''.join(random.choice('0123456789abcdef') for _iteration in range(64))
+
+            return stdout.strip()
+        except SubprocessError as ex:
+            display.error(ex.message)
+            display.warning('Failed to run docker image "{image}". Waiting a few seconds before trying again.')
+            docker_rm(args, name)  # podman doesn't remove containers after create if run fails
+            time.sleep(3)
+
+    raise ApplicationError(f'Failed to run docker image "{image}".')
+
+
+def start_container(args: EnvironmentConfig, container_id: str) -> tuple[t.Optional[str], t.Optional[str]]:
+    """Start a docker container by name or ID."""
+    options: list[str] = []
+
+    for _iteration in range(1, 3):
+        try:
+            return docker_start(args, container_id, options)
+        except SubprocessError as ex:
+            display.error(ex.message)
+            display.warning(f'Failed to start docker container "{container_id}". Waiting a few seconds before trying again.')
+            time.sleep(3)
+
+    raise ApplicationError(f'Failed to start docker container "{container_id}".')
+
+
+def get_container_ip_address(args: EnvironmentConfig, container: DockerInspect) -> t.Optional[str]:
+    """Return the IP address of the container for the preferred docker network."""
+    if container.networks:
+        network_name = get_docker_preferred_network_name(args)
+
+        if not network_name:
+            # Sort networks and use the first available.
+            # This assumes all containers will have access to the same networks.
+            network_name = sorted(container.networks.keys()).pop(0)
+
+        ipaddress = container.networks[network_name]['IPAddress']
+    else:
+        ipaddress = container.network_settings['IPAddress']
+
+    if not ipaddress:
+        return None
+
+    return ipaddress
+
+
+@mutex
+def get_docker_preferred_network_name(args: EnvironmentConfig) -> str | None:
```

Comment:
I'm surprised to see the `|` syntax here. I thought it appeared in Python 3.9/3.10. Don't we need to remain Python 3.8-compliant?

Reply:
This is okay. These files have `from __future__ import annotations` ([PEP 563](https://peps.python.org/pep-0563/)) that's available in 3.7+. This allows using the newer syntax on older Pythons when annotations aren't used in the runtime context.

FTR, the controller supports 3.9, 3.10, and 3.11 in 2.14 and devel, and the `|` syntax was added in Python 3.10, but that doesn't matter here as I said.

Reply:
> But I think I see your point — you imply that because the annotations are strings, they won't be evaluated in runtime, and it'll only have effect under things line MyPy.

Correct.

> I'm still not sure about this because I don't know if the lexer will be able to understand this syntax which is happening way earlier. I'm pretty sure it'd work if quoted explicitly, but I'm not confident about the unquoted annotations parsing.

This does indeed work properly with Mypy (as evidenced by the passing sanity tests), and it is not necessary to quote the annotations when the annotations when using `from __future__ import annotations`.

```
Python 3.9.15 (main, Nov 17 2022, 00:00:00) 
[GCC 12.2.1 20220819 (Red Hat 12.2.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> testvar: str|None = ""
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
>>> testvar: "str|None" = ""
>>> from __future__ import annotations
>>> testvar: str|None = ""
```

c.f. https://peps.python.org/pep-0585/#backwards-compatibility

Reply:
I've updated the type hints for consistency with the existing code. However, as @gotmax23 pointed out, they do work at runtime and in mypy due to the use of the annotations future. There are cases where it could be an issue, such as when the annotations are evaluated at runtime -- but ansible-test isn't doing that.